[Orch] Check cache for tags for v4 ecs

JLineaweaver opened the pull request:
<!--
* Contributors are encouraged to read our [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Please fill the below sections if possible with relevant information or links.
-->
### What does this PR do?
- Correctly respect the `ecs_collect_resource_tags_ec2` variable when calling the ECS Agent
- Start caching tags to reduce burden on the ECS Agent
- Start logging error responses from the ECS agent.

### Motivation
We are currently making too many requests for tags to the ECS Agent. We want to reduce the burden of total calls made. We also want to respect the config `ecs_collect_resource_tags_ec2` to turn off tagging calls. In addition we want to make sure errors are visible to the user when the ECS Agent returns errors.
### Describe how you validated your changes
I deployed this agent to an ECS cluster. Then I was able to see the calls change from getTaskWithTags to getTask after tags are stored in the cache.
<!--
Validate your changes before merge, ensuring that:
* Your PR is tested by static / unit / integrations / e2e tests
* Your PR description details which e2e tests cover your changes, if any
* The PR description contains details of how you validated your changes. If you validated changes manually and not through automated tests, add context on why automated tests did not fit your changes validation.

If you want additional validation by a second person, you can ask reviewers to do it. Describe how to set up an environment for manual tests in the PR description. Manual validation is expected to happen on every commit before merge.

Any manual validation step should then map to an automated test. Manual validation should not substitute automation, minus exceptions not supported by test tooling yet.
-->

### Possible Drawbacks / Trade-offs

### Additional Notes
<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

JLineaweaver committed:
[Orch] Check cache for tags for v4 ecs

JLineaweaver committed:
Add tags to the cache

JLineaweaver committed:
Change from entity to using task.Arn

JLineaweaver committed:
Return missing error

JLineaweaver committed:
Split call for getting tags or no tags

JLineaweaver committed:
Merge branch 'main' into jlineaweaver/v4-tag-cache

JLineaweaver committed:
Create map in collector test

JLineaweaver committed:
Try and fix assignment bug

JLineaweaver committed:
Try change of return call and add comments

JLineaweaver committed:
Delete extra err variable

JLineaweaver committed:
Add missing check for collectResourceTags

JLineaweaver committed:
Change notes and update test

JLineaweaver committed:
Refactor tests and check cache for tags

stzou commented on comp/core/workloadmeta/collectors/internal/ecs/v4parser.go:
```diff
@@ -81,18 +84,54 @@ func (c *collector) getTaskWithTagsFromV4Endpoint(ctx context.Context, task v1.T
+			taskWithTags.TaskTags = rt.tags
+		}
+		if len(rt.containerInstanceTags) == 0 {
+			taskWithTags.ContainerInstanceTags = rt.containerInstanceTags
```
were we previously not add container instance tags with the v4 parser?

JLineaweaver replied:
Pretty sure both were coming from GetTaskWithTags

kangyili commented on comp/core/workloadmeta/collectors/internal/ecs/v4parser.go:
```diff
@@ -81,18 +84,54 @@ func (c *collector) getTaskWithTagsFromV4Endpoint(ctx context.Context, task v1.T
+	var returnTask v3or4.Task
+	// If we have tags, are collecting tags and have zero tags present in the cache
+	// Call getTaskWithTags
+	if c.hasResourceTags && c.collectResourceTags && len(rt.tags) == 0 && len(rt.containerInstanceTags) == 0 {
```
With `c.collectResourceTags` , users now need to turn on two feature flags for ECS collection

- ecs_collect_resource_tags_ec2
- ecs_task_collection_enabled

JLineaweaver replied:
So I believe the two flags do something different. `ecs_task_collection_enabled` allows users to collect tasks and only if that is true does `ecs_collect_resource_tags_ec2`. I think this could allow us to turn `ecs_collect_resource_tags_ec2` on by default since it shouldn't matter if `ecs_task_collection_enabled` if false. What do you think?

kangyili commented on comp/core/workloadmeta/collectors/internal/ecs/v4parser.go:
```diff
@@ -81,18 +84,54 @@ func (c *collector) getTaskWithTagsFromV4Endpoint(ctx context.Context, task v1.T
+			return v1TaskToV4Task(task), err
+		}
+		// Add tags if they exist
+		if len(rt.tags) == 0 {
```
should be `if len(rt.tags) != 0 {` ?

JLineaweaver replied:
Good catch on this

JLineaweaver replied:
Added missing test and corrected test data

kangyili reviewed (approved):
Overall logic makes sense to me. Hitting the rate limit during the agent's first run seems unavoidable when the tag cache is still empty.

JLineaweaver committed:
Adress comment

JLineaweaver committed:
Add missing error fields to result and add test

Extra test checks to make sure tags are correctly set when the agent
calls get task after the cache has cleared

JLineaweaver committed:
Release notes

JLineaweaver committed:
Change release note length and line length

JLineaweaver committed:
Address linting issues

JLineaweaver committed:
Move releasenotes to correct folder

JLineaweaver committed:
Improve ecs log line

JLineaweaver committed:
Merge branch 'main' into jlineaweaver/v4-tag-cache

gabedos commented on comp/core/workloadmeta/collectors/internal/ecs/v4parser_test.go:
```diff
@@ -97,19 +109,96 @@ func TestPullWithTaskCollectionEnabledWithV4Parser(t *testing.T) {
+	store.notifiedEvents = store.notifiedEvents[:0]
+
+	// Second pull - should fetch from API again since cache was cleared
+	fmt.Println("Josh second run")
```
nit

gabedos commented on comp/core/workloadmeta/collectors/internal/ecs/v4parser_test.go:
```diff
@@ -97,19 +109,96 @@ func TestPullWithTaskCollectionEnabledWithV4Parser(t *testing.T) {
+	defer cleanup()
+
+	// First pull - should fetch tags from API
+	fmt.Println("Josh first run")
```
nit

JLineaweaver replied:
Forgot to push. Pushed up now

JLineaweaver committed:
Remove josh logs

JLineaweaver commented on comp/core/workloadmeta/collectors/internal/ecs/v4parser.go:
```diff
@@ -81,18 +84,61 @@ func (c *collector) getTaskWithTagsFromV4Endpoint(ctx context.Context, task v1.T
+			containerInstanceTags: taskWithTags.ContainerInstanceTags,
+		}
+
+		c.resourceTags[task.Arn] = rt
```
https://github.com/datadog/datadog-agent/blob/jlineaweaver%2Fv4-tag-cache/comp/core/workloadmeta/collectors/internal/ecs/ecs.go#L185-L188

I believe this is the code that cleans it up
